MFiles.VAF.Extensions/ExtensionMethods/MFSearchBuilderExtensionMethods/SegmentedSearch.cs

```diff
@@ -0,0 +1,253 @@
+		public const int DefaultNumberOfItemsInSegment = 1000;
+
+		/// <summary>
+		/// The maximum index segment (must be greater than zero).
+		/// </summary>
+		public const int DefaultMaximumSegmentIndex = 100001;
+
+		/// <summary>
+		/// The default search timeout in seconds.
+		/// </summary>
+		public const int DefaultSearchTimeoutInSeconds = 10;
+
+		/// <summary>
+		/// Executes a method for each <see cref="ObjVerEx"/> using a segmented search.
+		/// </summary>
+		/// <param name="builder">Search Builder to use for queries.</param>
+		/// <param name="objVerExDelegate">Method to execute for each object.</param>
+		/// <param name="startSegment">The (zero-based) index of the segment to start at.</param>
+		/// <param name="segmentLimit">The number of total segments to process. See <see cref="DefaultMaximumSegmentIndex"/>.</param>
+		/// <param name="segmentSize">The number of items to include in each segment. See <see cref="DefaultNumberOfItemsInSegment"/>.</param>
+		/// <param name="searchTimeoutInSeconds">The timeout for each search. See <see cref="DefaultSearchTimeoutInSeconds"/>.</param>
+		/// <returns>Total count of objects matching conditions.</returns>
+		/// <remarks>Note that <paramref name="searchTimeoutInSeconds"/> applies to the timeout on each segment search; if multiple segments are needed then the maximum time that this method takes to return will exceed the provided value.</remarks>
+		public static long ForEach(
+			this MFSearchBuilder builder,
+			Action<ObjVerEx> objVerExDelegate,
+			int startSegment = 0,
+			int segmentLimit = MFSearchBuilderExtensionMethods.DefaultMaximumSegmentIndex,
+			int segmentSize = MFSearchBuilderExtensionMethods.DefaultNumberOfItemsInSegment,
+			int searchTimeoutInSeconds = MFSearchBuilderExtensionMethods.DefaultSearchTimeoutInSeconds)
+		{
+			// Sanity checks
+			if (builder == null)
+				throw new ArgumentNullException(nameof(builder));
+			if (objVerExDelegate == null)
+				throw new ArgumentNullException(nameof(objVerExDelegate));
+			if (startSegment < 0)
+				throw new ArgumentOutOfRangeException(nameof(startSegment), "value must be greater than or equal to 0");
+			if (segmentLimit <= 0)
+				throw new ArgumentOutOfRangeException(nameof(segmentLimit), "value must be greater 0");
+			if (segmentSize <= 0)
+				throw new ArgumentOutOfRangeException(nameof(segmentSize), "value must be greater than 0");
+			if (searchTimeoutInSeconds < 0)
+				throw new ArgumentOutOfRangeException(nameof(searchTimeoutInSeconds), "The search timeout must be greater than zero, or zero to indicate an indefinite timeout");
+
+			int DefaultSearchHandler(Vault vault, SearchConditions searchConditions)
+			{
+				var newBuilder = new MFSearchBuilder(vault, searchConditions);
+				var searchResults = newBuilder.FindEx(searchTimeoutInSeconds: searchTimeoutInSeconds);
+
+				searchConditions.Remove(searchConditions.Count);
+
+				if (searchConditions.Count <= 0)

```

Comment:
If the search conditions are empty here then the processing of the search results will not be called.  Is this expected behaviour?

Reply:
I actually have no idea how that got there. I removed this. It would always succeed to remove 1 because it has the segment search condition, but it would be useless for an empty search.